Huawei: allow battery charge from PV in hold mode (BC)

CiNcH83 opened the pull request:
Fixes #24674

### !!!Breaking Changes!!!
- **`maxdischargepower` (`Maximum discharge power`) is now a required parameter**

### Functional Changes
- **battery can now charge from PV in `hold` mode**
`hold` has previously been a hack, using a forced discharge with discharge power set to 0 W, not allowing any charging in this mode.
I tried everything to also make `hold` watchdog supervised, but failed. Doing a forced charge with grid charging disabled resulted in the system prefer charging battery with PV surplus over serving house loads. So I ended up sacrificing watchdog for PV charging in `hold` mode.
- **optimized `charge` case**
Only rewrite the necessary register periodically.
- **updated documentation**

See [LUNA2000 Battery Control Documentation](https://docs.google.com/document/d/1pCopCJ3HsOCmYLwUJZjzcpnZSOeC5v6yDGFRf6hBEi8/edit?tab=t.0)

### Testing
- [x] `normal` with PV power > demand: battery charged from PV with excess power
- [x] `normal` with PV power < demand and battery not empty: battery discharged to home/vehicle
- [x] `hold` with battery not full and PV power > demand: battery charged from PV with excess power
- [x] `hold` with battery charged and PV power < demand: no discharging to home/vehicle
- [x] `charge` with enough PV power available to charge entirely from PV: battery charged from PV only
- [x] `charge` with not enough PV power available to charge entirely from PV: battery partly charged from PV and grid
- [x] `charge` with no PV power available: battery charged from grid
- [x] Each state transition: proper register values verified




State transitions have been tested via REST API:
```
curl -X 'POST' 'http://localhost:7070/api/batterymode/normal' -H 'accept: application/json' -d ''
while true; do curl -X 'POST' 'http://localhost:7070/api/batterymode/hold' -H 'accept: application/json' -d ''; sleep 30; done
while true; do curl -X 'POST' 'http://localhost:7070/api/batterymode/charge' -H 'accept: application/json' -d ''; sleep 30; done
```

### Considerations
Since active battery control has so far been handled via the watchdog supervised _Forcible Charge/Discharge_ registers (which do not allow for PV charging in `hold` mode), the registers for setting the maximum charge/discharge power during normal operation remained unaffected. With the new approach proposed in this PR, we use the standard battery discharge power register to set maximum discharge power to 0 W in `hold` mode, effectively leaving PV charging intact. The downside is that if evcc goes boom or register writes fail while transitioning back to `normal` mode, the battery discharge power may remain at 0 W.

### TODO
- [x] Remove defaults and instead use dynamic parameter reading (#25908) ... done in #27493
- [x] Re-test all battery states/state transitions

CiNcH83 committed:
Default max. charge/discharge power set to 2500W

CiNcH83 committed:
Updated documentation

CiNcH83 committed:
Battery can now charge from PV in hold mode

CiNcH83 committed:
fix Porcelain

CiNcH83 committed:
Merge remote-tracking branch 'remotes/upstream/master' into huawei_battctrl_v2

CiNcH83 committed:
Merge branch 'master' into huawei_battctrl_v2

CiNcH83 committed:
default maxdischargepower limited to 2200 W due to SUN2000-2KTL-L1

CiNcH83 commented:
Re-open for discussion.

/cc @andig

CiNcH83 committed:
set maxchargepower back to 10000 as force registers are not bounds-checked

mucki12 commented:
As I said — if that’s what’s wanted. Luckily, you can also integrate the battery as a custom device. That way, you get the more stable implementation.

BTW: my test was negative

andig commented:
> As I said — if that’s what’s wanted. Luckily, you can also integrate the battery as a custom device. That way, you get the more stable implementation.
> 
> BTW: my test was negative

Then the PR is misleading:

> Tested by @CiNcH83 & @mucki12 & @LotharWoman

CiNcH83 commented:
Your test has been negative because of [this](https://github.com/evcc-io/evcc/issues/25179). I re-opened this PR because the issue has been fixed by now.

mucki12 commented:
> Your test has been negative because of [this](https://github.com/evcc-io/evcc/issues/25179). 

Sorry, then I must have misunderstood the linked PR. I was having issues even though I didn’t make any external changes to the mode, and I thought the PR was intended to handle incorrect behavior caused by additional external switching via the API. But it doesn’t really matter — I’ll just integrate the LUNA as a custom battery, since, as I mentioned, the Huawei fallback in case of malfunction is much more important to me.

mucki12 commented:
Ok, but with this PR you still lose the Huawei fallback, right? Or did I miss something?

CiNcH83 commented on templates/definition/meter/huawei-sun2000.yaml:
```diff
@@ -32,8 +38,9 @@ params:
-    default: 10000
+    default: 2500
   - name: maxdischargepower
+    default: 2200
```
Can't figure out how this is supposed to be implemented.

andig replied:
> I now removed the default

You haven't? Unfortunately we also can't make these fields non-advanced :(

CiNcH83 commented:
The problem has been that _Maximum discharge power_ remained `0` after switching from `hold` to `normal`.

mucki12 commented:
During a test with the planned hold mode, the battery discharge power was once at 0, even though evcc was already back in normal mode. I can no longer trace how this happened. As a result, no battery lock was shown in evcc, but it was still effectively active. Possibly, the Huawei system “swallowed” the Modbus change when switching from hold to normal. That can happen, but as mentioned, it wouldn’t have been an issue with the old template.

CiNcH83 committed:
removed references to § 14a EnWG

CiNcH83 committed:
prettify links in doc

CiNcH83 commented:
Yes, this is in fact the last(?) open issue of this PR, see [above discussion](https://github.com/evcc-io/evcc/pull/25024#discussion_r2518126941).

2200 W would indeed be a bad default. Even though all configs would accept it, you will hardly find the very config where this is the optimal value (only for the 2 kW single-phase inverter).

LotharWoman commented:
> 3. Via ToU

Is no longer practical with the latest firmware. Enabling TOU will no longer be able to wake up the battery and inverter from hibernate.

mucki12 commented:
> Enabling TOU will no longer be able to wake up the battery and inverter from hibernate.

Correct – at least with the current firmware versions (although I still assume that Huawei will fix this later). However, in my view, this plays a rather minor role in hold mode. If the LUNA is sleeping, I don't want to wake it up because of 14a. But for charge mode, this definitely needs to be taken into account.

mucki12 commented:
Thank you!

That makes sense. Unfortunately, it must be noted that there is still no perfect alternative method. Control via discharge power is, of course, possible, but it still has significant disadvantages in terms of handling and stability.

LotharWoman commented:
I used the TOU method until last year. Also works very well.
 Once set up, it can also be used instead of “ForceCharge”.
 You can also easily customize it further by charging power and max. gridcharge SoC.
 The problem is that it makes things very complex. And the error rate is quite high if the Modbus data is not transmitted cleanly.
 I therefore still had a control instance in my automations that checked the status again and retransmitted if necessary.
  For nightly loading, I still use TOU in my automations.
 I still find “hold” by limiting the charging power to be the best and easiest solution during the day.
  If I remember correctly, the same problem existed with Growatt. There, too, “hold” is probably controlled via the discharge power. We should take a closer look at the templates of other systems.

CiNcH83 committed:
Merge remote-tracking branch 'remotes/upstream/master' into huawei_battctrl_v2

CiNcH83 committed:
maxdischargepower made required

andig commented:
Stuck in endless loop, we can ignore it. What‘s left to do here?

andig commented on templates/definition/meter/huawei-sun2000.yaml:
```diff
@@ -8,13 +8,17 @@ capabilities: ["battery-control"]
+
+      Modbus TCP erfordert die Freischaltung in den Kommunikationseinstellungen des Wechselrichters via "Installateur Zugang", siehe [Modbus TCP Aktivierungs-Anleitung](https://forum.huawei.com/enterprise/en/modbus-tcp-guide/thread/667250677153415168-667213868771979264).
+
+      Es ist zu beachten, dass die maximale Batterie-Entladeleistung standardmäßig auf 2200W eingestellt ist. Je nach installiertem Wechselrichter sowie der Anzahl der Batterietürme/Module kann der unterstützte Maximalwert deutlich höher liegen. Weitere Informationen können den Datenblättern von Huawei entnommen werden.
```
Das stimmt nicht mehr?

CiNcH83 commented:
I would like to re-test all battery states and state transitions. But I will need some sun and energy stored in the battery for that...

CiNcH83 committed:
updated documentation

CiNcH83 committed:
updated documentation

forceaccharging does not prevent standby anymore with newer firmware versions.

CiNcH83 committed:
Merge remote-tracking branch 'remotes/upstream/master' into huawei_battctrl_v2

CiNcH83 committed:
Merge branch 'huawei_battctrl_v2' of github.com:CiNcH83/evcc into huawei_battctrl_v2

CiNcH83 commented:
Closing this again. It is still problematic for the following reasons:

- no watchdog protection: SDongleA-05 is known to be problematic, especially certain dongle versions (non-004 PN with STM chip) in conjunction with [Home Assistant huawei_solar](https://github.com/wlcrs/huawei_solar) via modbus-proxy
- it may be problematic to rewrite static configuration registers (like `47077`) to some non-volatile flash memory every 15s (don't know whether only changes are effectively being written)

Currently checking with Huawei for an alternative solution.

CiNcH83 committed:
Merge remote-tracking branch 'remotes/upstream/master' into huawei_battctrl_v2

CiNcH83 committed:
removed forceaccharging

AC charging is a static configuration setting which is enabled by default. It does not prevent standby with later FW versions any longer.

CiNcH83 committed:
only charge mode is watchdog protected

CiNcH83 commented:
Re-opened

Applied the following changes:
- like [Sonnen](https://github.com/evcc-io/evcc/blob/0918c62b44ecb7f698a5ae946f04af247ce7779a/templates/definition/meter/sonnenbatterie_eco56.yaml#L69), only `charge` mode is supervised by watchdog, `hold` is one-shot  (see discussion [here](https://github.com/evcc-io/evcc/discussions/27469#discussioncomment-16635606))
- <del>deprecated `forceaccharging` (does not prevent standby anymore)</del>

YET UNTESTED!

CiNcH83 committed:
fixed template

CiNcH83 committed:
tryfix template

CiNcH83 committed:
fixed indentation

CiNcH83 commented:
@andig After removing `forceaccharging`, evcc does not boot anymore as it is still in the config. It can easily be removed from `evcc.yaml`, but how is this supposed to work with UI config?

andig commented:
Make it deprecated: true instead

CiNcH83 committed:
fixed charge mode

CiNcH83 committed:
make forceaccharging deprecated

CiNcH83 committed:
improved battery mode

CiNcH83 committed:
restored forceaccharging - it doesn't hurt

CiNcH83 committed:
fixed indentation

CiNcH83 committed:
also disable AC charging in hold

CiNcH83 commented:
@sourcery-ai dismiss

CiNcH83 committed:
Merge remote-tracking branch 'remotes/upstream/master' into huawei_battctrl_v2

CiNcH83 commented:
@sourcery-ai dismiss

CiNcH83 committed:
further optimized charge case - only 1 register written periodically

CiNcH83 committed:
updated documentation

CiNcH83 committed:
Merge remote-tracking branch 'remotes/upstream/master' into huawei_battctrl_v2

andig commented:
Nothing I could review. If it works it works. Merge?

CiNcH83 commented:
Tested it up and down the past two days and it works as intended.

I think it makes sense to pull `hold` out of the watchdog as there is no watchdog for it on device side. So periodically rewriting `47077` (`Max. discharge power [W]`) is for the fishes.

Also the other change with respect to force `charge` mode makes sense IMHO, only periodically rewriting the necessary registers via watchdog.

andig commented:
...and where does that leave us wrt the question of merging?

CiNcH83 commented:
That was a big YES from my side...

CiNcH83 commented:
The other thing is that previous `hold` was a hack. It was a forced discharge with discharge power set to 0 W, also not allowing any charging in `hold` mode.

I tried everything to also make `hold` watchdog supervised, but failed. Doing a forced charge with grid charging disabled resulted in the system prefering charging battery with PV surplus over serving house loads. So not a good solution either. So I ended up sacrificing watchdog for PV charging in `hold` mode.

hebe2k commented:
@CiNcH83 How is this supposed to work when using UI mode for configuration? I tried to update to evcc version 0.306.1 and now I get the error:

meter: [db:3] cannot create template 'db:3': cannot create meter type 'template:huawei-sun2000': missing required `maxdischargepower`

CiNcH83 commented:
You should still be able to enter the configuration page and enter the LUNA2000 settings. There you enter the `Maximum discharge power`.

hebe2k commented:
Thanks, I just had to enter the setup for the battery, save again and restart, then everything was fine.

CiNcH83 commented:
@andig When is the docu typically being updated? My changes do not seem to be in yet…

andig commented:
On release

CiNcH83 commented:
Hmmm.. it has been released. Refreshed the docu page multiple times...

black-snake commented:
Hi, just saw that this change broke my existing configuration since I don’t have `maxdischargepower` defined.

Now, I am wondering which value I have to enter there, given the attached specs of my battery? 5 kW (nominal/ permanent) discharge power or 7 kW (maximum/ short-term) discharge power? Can anyone elaborate on this?

<img width="2360" height="712" alt="IMG_0272" src="https://github.com/user-attachments/assets/24930573-9021-4e7d-8eb7-02d728788cc9" />



Thanks so much.

CiNcH83 commented:
When you focus the `Maximum discharge power` field, evcc is supposed to suggest what your Huawei system is capable of, see [this PR](https://github.com/evcc-io/evcc/pull/27493). It is in fact the nominal power.

TRON4R commented:
Danke für Deine beständigen Optimierungen, @CiNcH83 
Nur zu meinem Verständnis: welche Lösung wurde jetzt am Ende gewählt? Ich habe zwar versucht, die ganzen Diskussionen verteilt über die drei PRs zu verfolgen, aber bin am Ende nicht so 100%ig schlau draus geworden.

Wenn ich jetzt upgrade auf die 0.306.x, dann nehme ich jetzt das Risiko in Kauf, dass bei einem Absturz von evcc oder einem Modbus-Hänger meines SDongle zu einem ungünstigen Zeitpunkt am Ende ein Maximumdischarge-Wert im Huawei-System stehen bleibt, den ich eigentlich nicht haben möchte, richtig?

Wenn ich das nicht haben möchte, muss ich die LUNA2000 als Custom-Battery(?) selbst definieren statt das Huawei-Template auch dafür zu verwenden, richtig?

Dass evcc die Batterie auf `Hold` packt, ist ja ein relativ häufiger Fall, korrekt? Also z.B. wenn ich "Batterieunterstütztes Fahrzeugladen" auf 80% und 90% gesetzt habe, dann würde das auf `Hold` setzen solange geschehen, bis mein Hausakku 80% SoC erreicht hat. Und auch im Schnell-Modus und bei geplantem Laden wird ja nach meinem Verständnis dann `Hold` aktiviert, wenn der Button "Verhindere Entladung im Schnell-Modus und bei geplantem Laden." eingeschaltet ist. Also ist nach meinem Eindruck ein aktives `Hold` der deutlich häufigere Zustand, oder?

CiNcH83 commented:
Im Grunde alles richtig. Beim Verlassen vom `hold` Modus bleibt die `Max. Discharge Power` auf `0`, wenn der Übergang auf `normal` nicht korrekt funktioniert.

Ich kann schon nachvollziehen, wenn Huawei-Nutzer mit Erfahrung hier etwas geschädigt sind. Dennoch wäre es mal schön, das ganze mit aktuellen Gegebenheiten zu testen. Ich habe in der Doku noch ein paar Warnings hinzugefügt, siehe [hier](https://docs.evcc.io/docs/devices/meters#sun2000-hybrid-wechselrichter).

TRON4R commented:
> Ich kann schon nachvollziehen, wenn Huawei-Nutzer mit Erfahrung hier etwas geschädigt sind. Dennoch wäre es mal schön, das ganze mit aktuellen Gegebenheiten zu testen. Ich habe in der Doku noch ein paar Warnings hinzugefügt, siehe [hier](https://docs.evcc.io/docs/devices/meters#sun2000-hybrid-wechselrichter).

Die Warnung bezieht sich allerdings nur auf potentiell unterbrochene Modbus-Verbindungen. Welche Konsequenzen das haben kann (also dass die Hausbatterie dann faktisch nicht mehr zum Liefern des Hausstroms genutzt werden kann, weil sie im ungünstigen Fall bei 0 Watt hängen bleibt), geht daraus nicht hervor. Insofern wird die Warnung für meisten Leute (welcher normale User liest schon die Issues?) ohne Bedeutung bleiben. Und ich glaube, vielen ist auch nicht klar, wie wackelig der Modbus von ihrem SDongle in der Praxis leider ist.

Ich muss gestehen, ich verzichte aus diesem Grunde jetzt auch erstmal auf ein evcc-Update. 
Wenn ich nach peniblem Überschussladen meines Autos tagsüber dann irgendwann nachts zufällig feststelle, dass ich trotz voller LUNA meinen Haus- und Wärmepumpenstrom aus dem Grid bezogen habe, dann würde mich das schon sehr wurmen. Erst recht, wenn ich sowas erst einige Tage später entdecken sollte. 

Und jetzt im Kopf mitschleppen zu müssen, nach jedem evcc-Ladevorgang nochmal sicherheitshalber die LUNA-Settings zu überprüfen ist mir auch zuviel (unnötige) mental load, sorry.

Wenn ich mal irgendwann Zeit habe, mir eine HA-Automation zu basteln, die einen Alarm schmeißt, wenn nach Sonnenuntergang der LUNA-Output auf 0 Watt steht, dann sieht es vielleicht anders aus. (Wobei ich aus dem Stegreif gerade gar nicht weiß, ob ich das über die Huawei-Integration von HA leicht lösen kann oder auch noch anfangen muss, selbst den Modbus abzufragen)

Bis dahin oder bis hier eine für alle Huawei-Besitzer gleichermaßen gute Lösung gefunden wurde, verabschiede ich mich (mit Bedauern) von weiteren evcc-Updates. 

Bitte nicht falsch verstehen, ich verstehe die Motivation dahinter vollkommen: evcc für alle Hardware möglichst einheitlich agieren zu lassen. Aber für mich - und einige andere die hier in diesen Issues mitdiskutiert haben - ist der Nachteil massiv größer als der Vorteil. Für mich hat evcc bisher perfekt funktioniert. Jetzt habe ich ein schweres Risiko drin, welches mir alle Bemühungen, nur so wenig Netzstrom wie möglich zu verbrauchen, mit einem leider nicht so unwahrscheinlichen Fehler zerschießen kann. 

Trotzdem danke für Deine regelmäßige Pflege der Huawei-Komponente hier in evcc. Vielleicht findet sich ja irgendwann noch eine stabilere Lösung.

CiNcH83 commented:
Solange alle nur schreiben, dass es in der Theorie schiefgehen kann, hilft uns das nicht weiter. Für den Discharge Lock gibt es effektiv keinen Watchdog Schutz bei Huawei. Was wir hatten, war nur ein Hack, der andere Nachteile hatte.

Technisch ist das so jedenfalls die korrekte Lösung. Wenn man halt mit HA parallel drauflos ballert, ist Huawei schnell mal überfordert.

CiNcH83 commented:
@andig Macht evcc einen Retry, wenn ein Modbus-Write fehlschlägt?

andig commented:
i.A. ja!

TRON4R commented:
> Solange alle nur schreiben, dass es in der Theorie schiefgehen kann, hilft uns das nicht weiter. Für den Discharge Lock gibt es effektiv keinen Watchdog Schutz bei Huawei. Was wir hatten, war nur ein Hack, der andere Nachteile hatte.
> 
> Technisch ist das so jedenfalls die korrekte Lösung. Wenn man halt mit HA parallel drauflos ballert, ist Huawei schnell mal überfordert.

Verstehe ich, aber ich wollte diese Änderung nicht haben. Insofern wäre es m.E. sinnvoller, wenn das mal einer von den Nutzern austestet, die diese Codeänderung veranlasst haben.

CiNcH83 commented:
Ich habe hier 2 Tage lang manuell (per REST API) Modi kreuz und quer geswitcht und konnte kein Fehlverhalten feststellen. Hier läuft aber auch kein HA.

CiNcH83 commented:
> i.A. ja!

i.A.? Wenn über ein Template ein Register geschrieben wird?

CiNcH83 commented:
Scheinbar wird da kein Retry gemacht. Wäre evtl. was für das Modbus Plugin. IMHO wäre das die beste Lösung bei einem Modbus-Schluckauf.

andig commented:
> Scheinbar wird da kein Retry gemacht

Gerne Issue mit Logfile... Frage zu unklar...

CiNcH83 commented:
OK. Habe jetzt folgendes gefunden:
```
func Backoff() *backoff.ExponentialBackOff {
	return backoff.NewExponentialBackOff(backoff.WithInitialInterval(20*time.Millisecond), backoff.WithMaxElapsedTime(10*time.Second))
}
```

D.h. da wird im Fehlerfall mehrfach bis zu 10s lang probiert? Wird ein Fehler nur dann geloggt, wenn auch die Retries fehlgeschlagen sind? Irgendwie werde ich aus dem Code da nicht so richtig schlau.

Ich versuche halt irgendwie, die Probleme mit Huawei zu lösen. Aber so geht das halt nicht, wenn die Beteiligten auf alten Sachen herumreiten und auch nicht bereit sind, den Problemen auf den Grund zu gehen. Ich weiß nur aus der Theorie, dass Wechselrichter-Kaskade und ein parallel auf das System ballerndes Home Assistant (meist in Kombination?) wohl problematisch sind. Ich habe keines von beidem...

andig commented:
> Aber so geht das halt nicht, wenn die Beteiligten auf alten Sachen herumreiten und auch nicht bereit sind, den Problemen auf den Grund zu gehen.

Probleme? https://github.com/evcc-io/evcc/pull/25024#issuecomment-4391602724

CiNcH83 commented:
OK. Beenden wir das.

Der Ansatz in diesem PR ist korrekt. Davon bin ich überzeugt. Ich habe das ausgiebig getestet. Solange das niemand mit komplexeren Systemen ausprobieren will, kommen wir eh nicht weiter. Meines ist mit nur einem Wechselrichter und ohne [huawei_solar](https://github.com/wlcrs/huawei_solar) sehr einfach. Und ja, ein HW-seitiger Watchdog wäre wünschenswert, aber auch ein Retry sollte einen Modbus Schluckauf lösen können.

Falls es doch jemand testen und konstruktiv mitarbeiten will...

- aktuelle FW-Stände wären wünschenswert
- manuell testen kann man die Umschaltung relativ einfach über das [REST-API](https://docs.evcc.io/docs/integrations/rest-api) - `batterymode` (konnte man da früher nicht mal die IP der eigenen evcc Instanz eingeben?)
```
curl -X 'POST' 'http://localhost:7070/api/batterymode/normal' -H 'accept: application/json' -d ''
while true; do curl -X 'POST' 'http://localhost:7070/api/batterymode/hold' -H 'accept: application/json' -d ''; sleep 30; done
while true; do curl -X 'POST' 'http://localhost:7070/api/batterymode/charge' -H 'accept: application/json' -d ''; sleep 30; done
```

andig commented:
> OK. Beenden wir das.

Die Frage stellt sich ja eh nicht- der PR ist drin! Wenns jetzt konkrete Probleme damit gibt brauchts ein Issue. Zumindest ist ja in den letzten 2 Wochen die Welt damit nicht untergegangen!

torwag commented:
3 WR in Kaskade mit jeweils 3 Luna Batterien, 1X 22KW openWB, 1 dezidierten Home-Assistant fürs Energie-Management, der wegen EVCC aktuell die  Huawei PV gerade nicht abfragt, da der modbusproxy nie so richtig wollte. WR sind über ihre Wifi-APs mit 3 Wifi USB Dongle am Server angeschlossen. EVCC und HA laufen in Containern auf einem proxmox host. Netzwerke werden über bridges in proxmox an den EVCC container weitergereicht. Als EAuto kommt die Ladezicke Renault Zoe zum Einsatz. 

Daily usage kommt von der holden Gattin, wenn das Auto nicht lädt, bin ich schnell auf ganz dünnen Eis unterwegs. 

Komplex genug? 

Dann würde ich am liebsten nen Chat oder sogar ein Telefonat vorschlagen, dann könnte man sich einmal austauschen was ich testen kann und wie komplex das Ganze wird. Anschließend könnte ich mir dafür ein Zeitfenster freimachen. Wenns räumlich passt könnte man sich auch treffen, passende Hopfenkaltschorle soll ja beim debuggen helfen.

andig commented:
https://github.com/evcc-io/evcc/pull/25024#issuecomment-4396209129

TRON4R commented:
> 3 WR in Kaskade mit jeweils 3 Luna Batterien, 1X 22KW openWB, 1 dezidierten Home-Assistant fürs Energie-Management, der wegen EVCC aktuell die Huawei PV gerade nicht abfragt, da der modbusproxy nie so richtig wollte. WR sind über ihre Wifi-APs mit 3 Wifi USB Dongle am Server angeschlossen. 

So ein Setup mit 3 WRs, die dann auch noch über 3 WiFi Dongle angeschlossen sind statt uber einen gemeinsamen SDongle ist aber schon einr sehr ungewöhnliche Konfiguration.

andig commented:
Es dürfte jetzt klar sein, dass das vermeintliche Problem sich nicht hier im PR diskutieren lässt, daher geschlossen. 

@torwag Ansonsten gilt weiterhin https://github.com/evcc-io/evcc/pull/25024#issuecomment-4396209129 (wie jetzt 2x wiederholt!).
